### Importação das bibliotecas

In [0]:

from pyspark.sql import SparkSession, DataFrame
from pyspark.sql.functions import col, sum, regexp_replace, when, regexp_replace,  year, expr


In [0]:

spark = SparkSession.builder \
    .appName("Análise de Dados Database Type Ecommerce UI Bakery") \
    .getOrCreate()

### Funções

In [0]:
def exibir_info_df(df: DataFrame) -> None:
    """
    Imprime o número de linhas, número de colunas e o esquema do DataFrame fornecido.

    Parâmetros:
        df (DataFrame): O DataFrame a ser analisado.

    Returns:
        None
    """

    print(f'\nLinhas = {df.count()} \nColunas = {len(df.columns)}\n')
    df.printSchema()


### Perguntas

####  - Qual país possui a maior quantidade de itens cancelados?

Dataframes abordados para essa questão:
- Orders
- Orders Details
- Customers

In [0]:
df_orders = spark.read.format('delta').load('/join/silver/df_orders')
exibir_info_df(df_orders)

df_orderdetails = spark.read.format('delta').load('/join/silver/df_orderdetails')
exibir_info_df(df_orderdetails)

df_customers = spark.read.format('delta').load('/join/silver/df_customers')
exibir_info_df(df_customers)


Linhas = 329 
Colunas = 7

root
 |-- order_number: integer (nullable = true)
 |-- order_date: date (nullable = true)
 |-- required_date: date (nullable = true)
 |-- shipped_date: date (nullable = true)
 |-- status: string (nullable = true)
 |-- comments: string (nullable = true)
 |-- customer_number: integer (nullable = true)


Linhas = 2997 
Colunas = 5

root
 |-- order_number: integer (nullable = true)
 |-- product_code: string (nullable = true)
 |-- quantity_ordered: integer (nullable = true)
 |-- price_each: decimal(10,2) (nullable = true)
 |-- order_line_number: short (nullable = true)


Linhas = 122 
Colunas = 12

root
 |-- customer_number: integer (nullable = true)
 |-- customer_name: string (nullable = true)
 |-- contact_last_name: string (nullable = true)
 |-- contact_first_name: string (nullable = true)
 |-- phone: string (nullable = true)
 |-- address_line1: string (nullable = true)
 |-- city: string (nullable = true)
 |-- state: string (nullable = true)
 |-- postal_code: s

In [0]:
df_orders.filter('status == "Cancelled"').display()

order_number order_date required_date shipped_date status comments customer_number 10167 2003-10-23 2003-10-30 2023-01-01 Cancelled Customer called to cancel. The warehouse was notified in time and the order didn't ship. They have a new VP of Sales and are shifting their sales model. Our VP of Sales should contact them. 448 10179 2003-11-11 2003-11-17 2003-11-13 Cancelled Customer cancelled due to urgent budgeting issues. Must be cautious when dealing with them in the future. Since order shipped already we must discuss who would cover the shipping charges. 496 10248 2004-05-07 2004-05-14 2023-01-01 Cancelled Order was mistakenly placed. The warehouse noticed the lack of documentation. 131 10253 2004-06-01 2004-06-09 2004-06-02 Cancelled Customer disputed the order and we agreed to cancel it. We must be more cautions with this customer going forward, since they are very hard to please. We must cover the shipping fees. 201 10260 2004-06-16 2004-06-22 2023-01-01 Cancelled Customer heard complaints from their customers and called to cancel this order. Will notify the Sales Manager. 357 10262 2004-06-24 2004-07-01 2023-01-01 Cancelled This customer found a better offer from one of our competitors. Will call back to renegotiate. 141

In [0]:
orders_cancelled = df_orders.filter('status == "Cancelled"')
orders_cancelled.display()

order_number order_date required_date shipped_date status comments customer_number 10167 2003-10-23 2003-10-30 2023-01-01 Cancelled Customer called to cancel. The warehouse was notified in time and the order didn't ship. They have a new VP of Sales and are shifting their sales model. Our VP of Sales should contact them. 448 10179 2003-11-11 2003-11-17 2003-11-13 Cancelled Customer cancelled due to urgent budgeting issues. Must be cautious when dealing with them in the future. Since order shipped already we must discuss who would cover the shipping charges. 496 10248 2004-05-07 2004-05-14 2023-01-01 Cancelled Order was mistakenly placed. The warehouse noticed the lack of documentation. 131 10253 2004-06-01 2004-06-09 2004-06-02 Cancelled Customer disputed the order and we agreed to cancel it. We must be more cautions with this customer going forward, since they are very hard to please. We must cover the shipping fees. 201 10260 2004-06-16 2004-06-22 2023-01-01 Cancelled Customer heard complaints from their customers and called to cancel this order. Will notify the Sales Manager. 357 10262 2004-06-24 2004-07-01 2023-01-01 Cancelled This customer found a better offer from one of our competitors. Will call back to renegotiate. 141

- Solução utilizando SQL

Criando tabelas temporárias

In [0]:

orders_cancelled.createOrReplaceTempView('cancel')
df_orderdetails.createOrReplaceTempView('details')
df_customers.createOrReplaceTempView('customer')

In [0]:
%sql

SELECT 
  c.order_number,
  customer.customer_number,
  customer.country,
  SUM(d.quantity_ordered) AS cont_quantity_ordered
FROM details AS d
INNER JOIN cancel AS c
INNER JOIN customer
ON d.order_number = c.order_number AND customer.customer_number = c.customer_number
GROUP BY c.order_number, customer.customer_number, customer.country
ORDER BY cont_quantity_ordered DESC 

order_number customer_number country cont_quantity_ordered 10262 141 Spain 605 10167 448 Sweden 550 10248 131 USA 454 10253 201 UK 429 10260 357 New Zealand 308 10179 496 New Zealand 288

In [0]:

%sql

WITH cte_tabela AS (
  SELECT 
    c.order_number,
    cl.customer_number,
    cl.country,
    SUM(d.quantity_ordered) AS cont_quantity_ordered
  FROM details AS d
  INNER JOIN cancel AS c
  INNER JOIN customer AS cl
  ON d.order_number = c.order_number AND cl.customer_number = c.customer_number
  GROUP BY c.order_number, cl.customer_number, cl.country
  ORDER BY cont_quantity_ordered DESC 
)
SELECT 
   country AS pais,
   SUM(cont_quantity_ordered) AS contagem_itens_cancelados
FROM cte_tabela
GROUP BY pais
ORDER BY contagem_itens_cancelados DESC

pais contagem_itens_cancelados Spain 605 New Zealand 596 Sweden 550 USA 454 UK 429

In [0]:

%sql

WITH cte_tabela AS (
  SELECT 
    c.order_number,
    cl.customer_number,
    cl.country,
    SUM(d.quantity_ordered) AS cont_quantity_ordered
  FROM details AS d
  INNER JOIN cancel AS c
  INNER JOIN customer AS cl
  ON d.order_number = c.order_number AND cl.customer_number = c.customer_number
  GROUP BY c.order_number, cl.customer_number, cl.country
  ORDER BY cont_quantity_ordered DESC 
)
SELECT 
   country AS pais,
   SUM(cont_quantity_ordered) AS contagem_itens_cancelados
FROM cte_tabela
GROUP BY pais
ORDER BY contagem_itens_cancelados DESC

pais contagem_itens_cancelados Spain 605 New Zealand 596 Sweden 550 USA 454 UK 429

Databricks visualization. Run in Databricks to view.

Consulta final apresentando o resultado

In [0]:
%sql

WITH cte_tabela AS (
  SELECT 
    c.order_number,
    cl.customer_number,
    cl.country,
    SUM(d.quantity_ordered) AS cont_quantity_ordered
  FROM details AS d
  INNER JOIN cancel AS c
  INNER JOIN customer AS cl
  ON d.order_number = c.order_number AND cl.customer_number = c.customer_number
  GROUP BY c.order_number, cl.customer_number, cl.country
  ORDER BY cont_quantity_ordered DESC 
)
SELECT 
   country AS pais,
   SUM(cont_quantity_ordered) AS contagem_itens_cancelados
FROM cte_tabela
GROUP BY pais
ORDER BY contagem_itens_cancelados DESC
LIMIT 1

pais contagem_itens_cancelados Spain 605

- Solução utilizando PySpark

In [0]:
df_q1 = orders_cancelled.join(df_orderdetails, orders_cancelled.order_number == df_orderdetails.order_number, 'INNER')\
                             .join(df_customers, df_customers.customer_number == orders_cancelled.customer_number, 'INNER')\
                             .select(df_customers.country, df_orderdetails.quantity_ordered)

In [0]:
df_q1.display()

country quantity_ordered Sweden 44 Sweden 43 Sweden 46 Sweden 34 Sweden 33 Sweden 21 Sweden 20 Sweden 32 Sweden 29 Sweden 43 Sweden 29 Sweden 46 Sweden 24 Sweden 28 Sweden 40 Sweden 38 New Zealand 24 New Zealand 47 New Zealand 27 New Zealand 45 New Zealand 24 New Zealand 34 New Zealand 23 New Zealand 25 New Zealand 39 USA 20 USA 21 USA 32 USA 42 USA 42 USA 48 USA 30 USA 23 USA 36 USA 40 USA 32 USA 30 USA 35 USA 23 UK 24 UK 22 UK 25 UK 41 UK 26 UK 24 UK 23 UK 33 UK 37 UK 40 UK 31 UK 40 UK 24 UK 39 New Zealand 46 New Zealand 30 New Zealand 44 New Zealand 32 New Zealand 29 New Zealand 23 New Zealand 23 New Zealand 27 New Zealand 21 New Zealand 33 Spain 49 Spain 32 Spain 34 Spain 34 Spain 24 Spain 46 Spain 49 Spain 48 Spain 40 Spain 49 Spain 40 Spain 44 Spain 33 Spain 27 Spain 35 Spain 21

In [0]:
df_q1.groupBy('country')\
     .agg(sum('quantity_ordered').alias('qtd_itens_devolvidos'))\
     .orderBy(col('qtd_itens_devolvidos').desc())\
     .display()

country qtd_itens_devolvidos Spain 605 New Zealand 596 Sweden 550 USA 454 UK 429

In [0]:
df_q1 = df_q1.groupBy('country')\
             .agg(sum('quantity_ordered').alias('qtd_itens_devolvidos'))\
             .orderBy(col('qtd_itens_devolvidos').desc())\
             .limit(1)

In [0]:
df_q1.display()

country qtd_itens_devolvidos Spain 605

Salvando o resultado

In [0]:
df_q1.write.mode('overwrite').format('delta').option('mergeSchema', 'True').save('/join/gold/resposta_q1')

> <b>Resposta</b>
-  O país com maior quantidade de itens cancelados é a Espanha com 605 itens devolvidos

#### - Qual o faturamento da linha de produto mais vendido, considere os itens com status 'Shipped', cujo o pedido foi realizado no ano de 2005

Dataframes abordados para essa questão:
- Orders
- Orders Details
- Products

In [0]:
df_orders = spark.read.format('delta').load('/join/silver/df_orders')
exibir_info_df(df_orders)

df_orderdetails = spark.read.format('delta').load('/join/silver/df_orderdetails')
exibir_info_df(df_orderdetails)

df_products = spark.read.format('delta').load('/join/silver/df_products')
exibir_info_df(df_products)


Linhas = 329 
Colunas = 7

root
 |-- order_number: integer (nullable = true)
 |-- order_date: date (nullable = true)
 |-- required_date: date (nullable = true)
 |-- shipped_date: date (nullable = true)
 |-- status: string (nullable = true)
 |-- comments: string (nullable = true)
 |-- customer_number: integer (nullable = true)


Linhas = 2997 
Colunas = 5

root
 |-- order_number: integer (nullable = true)
 |-- product_code: string (nullable = true)
 |-- quantity_ordered: integer (nullable = true)
 |-- price_each: decimal(10,2) (nullable = true)
 |-- order_line_number: short (nullable = true)


Linhas = 110 
Colunas = 9

root
 |-- product_code: string (nullable = true)
 |-- product_name: string (nullable = true)
 |-- product_line: string (nullable = true)
 |-- product_scale: string (nullable = true)
 |-- product_vendor: string (nullable = true)
 |-- product_description: string (nullable = true)
 |-- quantity_in_stock: short (nullable = true)
 |-- buy_price: decimal(10,2) (nullable = tru

In [0]:
df_orders.filter(expr('status = "Shipped" AND year(shipped_date) = 2005')).display()

order_number order_date required_date shipped_date status comments customer_number 10362 2005-01-05 2005-01-16 2005-01-10 Shipped No comments 161 10363 2005-01-06 2005-01-12 2005-01-10 Shipped No comments 334 10364 2005-01-06 2005-01-17 2005-01-09 Shipped No comments 350 10365 2005-01-07 2005-01-18 2005-01-11 Shipped No comments 320 10366 2005-01-10 2005-01-19 2005-01-12 Shipped No comments 381 10368 2005-01-19 2005-01-27 2005-01-24 Shipped Can we renegotiate this one? 124 10369 2005-01-20 2005-01-28 2005-01-24 Shipped No comments 379 10370 2005-01-20 2005-02-01 2005-01-25 Shipped No comments 276 10371 2005-01-23 2005-02-03 2005-01-25 Shipped No comments 124 10372 2005-01-26 2005-02-05 2005-01-28 Shipped No comments 398 10373 2005-01-31 2005-02-08 2005-02-06 Shipped No comments 311 10374 2005-02-02 2005-02-09 2005-02-03 Shipped No comments 333 10375 2005-02-03 2005-02-10 2005-02-06 Shipped No comments 119 10376 2005-02-08 2005-02-18 2005-02-13 Shipped No comments 219 10377 2005-02-09 2005-02-21 2005-02-12 Shipped Cautious optimism. We have happy customers here, if we can keep them well stocked. I need all the information I can get on the planned shippments of Porches 186 10378 2005-02-10 2005-02-18 2005-02-11 Shipped No comments 141 10379 2005-02-10 2005-02-18 2005-02-11 Shipped No comments 141 10380 2005-02-16 2005-02-24 2005-02-18 Shipped No comments 141 10381 2005-02-17 2005-02-25 2005-02-18 Shipped No comments 321 10382 2005-02-17 2005-02-23 2005-02-18 Shipped Custom shipping instructions sent to warehouse 124 10383 2005-02-22 2005-03-02 2005-02-25 Shipped No comments 141 10384 2005-02-23 2005-03-06 2005-02-27 Shipped No comments 321 10385 2005-02-28 2005-03-09 2005-03-01 Shipped No comments 124 10387 2005-03-02 2005-03-09 2005-03-06 Shipped We need to keep in close contact with their Marketing VP. He is the decision maker for all their purchases. 148 10388 2005-03-03 2005-03-11 2005-03-09 Shipped No comments 462 10389 2005-03-03 2005-03-09 2005-03-08 Shipped No comments 448 10390 2005-03-04 2005-03-11 2005-03-07 Shipped They want to reevaluate their terms agreement with Finance. 124 10391 2005-03-09 2005-03-20 2005-03-15 Shipped No comments 276 10392 2005-03-10 2005-03-18 2005-03-12 Shipped No comments 452 10393 2005-03-11 2005-03-22 2005-03-14 Shipped They want to reevaluate their terms agreement with Finance. 323 10394 2005-03-15 2005-03-25 2005-03-19 Shipped No comments 141 10395 2005-03-17 2005-03-24 2005-03-23 Shipped We must be cautions with this customer. Their VP of Sales resigned. Company may be heading down. 250 10396 2005-03-23 2005-04-02 2005-03-28 Shipped No comments 124 10397 2005-03-28 2005-04-09 2005-04-01 Shipped No comments 242 10398 2005-03-30 2005-04-09 2005-03-31 Shipped No comments 353 10399 2005-04-01 2005-04-12 2005-04-03 Shipped No comments 496 10400 2005-04-01 2005-04-11 2005-04-04 Shipped Customer requested that DHL is used for this shipping 450 10402 2005-04-07 2005-04-14 2005-04-12 Shipped No comments 406 10403 2005-04-08 2005-04-18 2005-04-11 Shipped No comments 201 10404 2005-04-08 2005-04-14 2005-04-11 Shipped No comments 323 10405 2005-04-14 2005-04-24 2005-04-20 Shipped No comments 209 10408 2005-04-22 2005-04-29 2005-04-27 Shipped No comments 398 10409 2005-04-23 2005-05-05 2005-04-24 Shipped No comments 166 10410 2005-04-29 2005-05-10 2005-04-30 Shipped No comments 357 10411 2005-05-01 2005-05-08 2005-05-06 Shipped No comments 233 10412 2005-05-03 2005-05-13 2005-05-05 Shipped No comments 141 10413 2005-05-05 2005-05-14 2005-05-09 Shipped Customer requested that DHL is used for this shipping 175 10416 2005-05-10 2005-05-16 2005-05-14 Shipped No comments 386 10418 2005-05-16 2005-05-24 2005-05-20 Shipped No comments 412 10419 2005-05-17 2005-05-28 2005-05-19 Shipped No comments 382

In [0]:
df_orders_q2 = df_orders.filter(expr('status = "Shipped" AND year(shipped_date) = 2005'))
df_orders_q2.display()

order_number order_date required_date shipped_date status comments customer_number 10362 2005-01-05 2005-01-16 2005-01-10 Shipped No comments 161 10363 2005-01-06 2005-01-12 2005-01-10 Shipped No comments 334 10364 2005-01-06 2005-01-17 2005-01-09 Shipped No comments 350 10365 2005-01-07 2005-01-18 2005-01-11 Shipped No comments 320 10366 2005-01-10 2005-01-19 2005-01-12 Shipped No comments 381 10368 2005-01-19 2005-01-27 2005-01-24 Shipped Can we renegotiate this one? 124 10369 2005-01-20 2005-01-28 2005-01-24 Shipped No comments 379 10370 2005-01-20 2005-02-01 2005-01-25 Shipped No comments 276 10371 2005-01-23 2005-02-03 2005-01-25 Shipped No comments 124 10372 2005-01-26 2005-02-05 2005-01-28 Shipped No comments 398 10373 2005-01-31 2005-02-08 2005-02-06 Shipped No comments 311 10374 2005-02-02 2005-02-09 2005-02-03 Shipped No comments 333 10375 2005-02-03 2005-02-10 2005-02-06 Shipped No comments 119 10376 2005-02-08 2005-02-18 2005-02-13 Shipped No comments 219 10377 2005-02-09 2005-02-21 2005-02-12 Shipped Cautious optimism. We have happy customers here, if we can keep them well stocked. I need all the information I can get on the planned shippments of Porches 186 10378 2005-02-10 2005-02-18 2005-02-11 Shipped No comments 141 10379 2005-02-10 2005-02-18 2005-02-11 Shipped No comments 141 10380 2005-02-16 2005-02-24 2005-02-18 Shipped No comments 141 10381 2005-02-17 2005-02-25 2005-02-18 Shipped No comments 321 10382 2005-02-17 2005-02-23 2005-02-18 Shipped Custom shipping instructions sent to warehouse 124 10383 2005-02-22 2005-03-02 2005-02-25 Shipped No comments 141 10384 2005-02-23 2005-03-06 2005-02-27 Shipped No comments 321 10385 2005-02-28 2005-03-09 2005-03-01 Shipped No comments 124 10387 2005-03-02 2005-03-09 2005-03-06 Shipped We need to keep in close contact with their Marketing VP. He is the decision maker for all their purchases. 148 10388 2005-03-03 2005-03-11 2005-03-09 Shipped No comments 462 10389 2005-03-03 2005-03-09 2005-03-08 Shipped No comments 448 10390 2005-03-04 2005-03-11 2005-03-07 Shipped They want to reevaluate their terms agreement with Finance. 124 10391 2005-03-09 2005-03-20 2005-03-15 Shipped No comments 276 10392 2005-03-10 2005-03-18 2005-03-12 Shipped No comments 452 10393 2005-03-11 2005-03-22 2005-03-14 Shipped They want to reevaluate their terms agreement with Finance. 323 10394 2005-03-15 2005-03-25 2005-03-19 Shipped No comments 141 10395 2005-03-17 2005-03-24 2005-03-23 Shipped We must be cautions with this customer. Their VP of Sales resigned. Company may be heading down. 250 10396 2005-03-23 2005-04-02 2005-03-28 Shipped No comments 124 10397 2005-03-28 2005-04-09 2005-04-01 Shipped No comments 242 10398 2005-03-30 2005-04-09 2005-03-31 Shipped No comments 353 10399 2005-04-01 2005-04-12 2005-04-03 Shipped No comments 496 10400 2005-04-01 2005-04-11 2005-04-04 Shipped Customer requested that DHL is used for this shipping 450 10402 2005-04-07 2005-04-14 2005-04-12 Shipped No comments 406 10403 2005-04-08 2005-04-18 2005-04-11 Shipped No comments 201 10404 2005-04-08 2005-04-14 2005-04-11 Shipped No comments 323 10405 2005-04-14 2005-04-24 2005-04-20 Shipped No comments 209 10408 2005-04-22 2005-04-29 2005-04-27 Shipped No comments 398 10409 2005-04-23 2005-05-05 2005-04-24 Shipped No comments 166 10410 2005-04-29 2005-05-10 2005-04-30 Shipped No comments 357 10411 2005-05-01 2005-05-08 2005-05-06 Shipped No comments 233 10412 2005-05-03 2005-05-13 2005-05-05 Shipped No comments 141 10413 2005-05-05 2005-05-14 2005-05-09 Shipped Customer requested that DHL is used for this shipping 175 10416 2005-05-10 2005-05-16 2005-05-14 Shipped No comments 386 10418 2005-05-16 2005-05-24 2005-05-20 Shipped No comments 412 10419 2005-05-17 2005-05-28 2005-05-19 Shipped No comments 382

- Solução utilizando SQL

Criando tabelas temporárias

In [0]:
df_orders_q2.createOrReplaceTempView('ordersq2')
df_orderdetails.createOrReplaceTempView('details')
df_products.createOrReplaceTempView('products')

In [0]:
%sql
WITH cte_tabela AS (
  SELECT 
    * 
  FROM details AS d
  INNER JOIN ordersq2 AS o 
  ON d.order_number = o.order_number
)
SELECT 
  p.product_line,
  SUM(c.price_each) AS faturamento
FROM cte_tabela AS c
INNER JOIN products AS p
ON c.product_code = p.product_code
GROUP BY p.product_line
ORDER BY faturamento DESC
-- LIMIT(1)

product_line faturamento Classic Cars 15559.72 Vintage Cars 5861.56 Motorcycles 5488.50 Trucks and Buses 4073.94 Planes 3165.93 Ships 1746.82 Trains 705.81

In [0]:
%sql
WITH cte_tabela AS (
  SELECT 
    * 
  FROM details AS d
  INNER JOIN ordersq2 AS o 
  ON d.order_number = o.order_number
)
SELECT 
  p.product_line,
  SUM(c.price_each) AS faturamento
FROM cte_tabela AS c
INNER JOIN products AS p
ON c.product_code = p.product_code
GROUP BY p.product_line
ORDER BY faturamento DESC


product_line faturamento Classic Cars 15559.72 Vintage Cars 5861.56 Motorcycles 5488.50 Trucks and Buses 4073.94 Planes 3165.93 Ships 1746.82 Trains 705.81

Databricks visualization. Run in Databricks to view.

In [0]:
%sql
WITH cte_tabela AS (
  SELECT 
    * 
  FROM details AS d
  INNER JOIN ordersq2 AS o 
  ON d.order_number = o.order_number
)
SELECT 
  p.product_line,
  SUM(c.price_each) AS faturamento
FROM cte_tabela AS c
INNER JOIN products AS p
ON c.product_code = p.product_code
GROUP BY p.product_line
ORDER BY faturamento DESC
LIMIT(1)

product_line faturamento Classic Cars 15559.72

- Solução utilizando PySpark

In [0]:
df_orderdetails.join(df_orders_q2, df_orderdetails.order_number == df_orders_q2.order_number, 'INNER')\
               .select(df_orderdetails.price_each, df_orderdetails.product_code).display()

price_each product_code 182.04 S10_4698 131.04 S12_2823 53.91 S18_2625 91.29 S24_1578 180.95 S12_1099 106.87 S12_3380 68.63 S12_3990 103.64 S12_4675 61.60 S18_1889 69.15 S18_3278 124.94 S18_3482 52.22 S18_3782 123.50 S18_4721 70.08 S24_2000 56.10 S24_2360 52.05 S24_3371 113.75 S24_3856 75.99 S24_4620 92.90 S32_1374 38.22 S32_2206 116.06 S18_1129 82.66 S32_4485 68.34 S50_4713 116.65 S18_1984 105.60 S18_2870 154.10 S18_3232 73.60 S24_2766 115.09 S24_2887 83.04 S24_3191 93.16 S24_3432 36.52 S24_3969 195.01 S10_1949 89.38 S18_1342 46.36 S18_1367 100.30 S18_2949 51.84 S18_2957 90.06 S18_3136 80.36 S18_3320 93.49 S24_4258 128.53 S10_4962 128.47 S12_1666 100.34 S18_1097 101.87 S18_2319 60.16 S18_2432 167.65 S18_3232 105.34 S18_4600 41.76 S18_4668 63.99 S32_3522 178.71 S12_1108 104.28 S12_4473 160.46 S18_2238 53.75 S24_1444 126.51 S24_2300 35.01 S24_2840 95.81 S24_4048 82.83 S32_1268 44.37 S32_2509 53.44 S32_3207 97.23 S50_1392 83.95 S700_2824 146.55 S12_3148 140.15 S12_3891 131.13 S18_3140 91.76 S18_3259 119.20 S18_4027 78.99 S18_4522 102.00 S24_2011 56.82 S50_1514 74.48 S700_1938 118.32 S10_4757 143.50 S18_1662 75.70 S18_3029 99.52 S18_3856 58.92 S24_2841 82.31 S24_3151 53.92 S24_3420 83.86 S24_3816 62.10 S24_3949 58.00 S700_1138 76.94 S700_2047 69.39 S700_2610 94.16 S700_3505 83.42 S700_3962 68.11 S700_4002 44.20 S72_1253 48.05 S72_3212 115.37 S10_2016 158.80 S10_4698 75.19 S18_2581 48.46 S18_2625 112.70 S24_1578 107.23 S24_1785 76.56 S10_1678 184.84 S12_1099 150.62 S12_2823 67.03 S24_2000 60.26 S24_2360 60.13 S24_4278 87.90 S32_1374 59.85 S32_4289 96.95 S32_4485 36.22 S50_1341 69.16 S50_4713 86.77 S700_1691 94.73 S700_2466 98.48 S700_2834 69.60 S700_3167 98.65 S12_3380 65.44 S12_3990 112.86 S12_4675 124.56 S18_1129 61.60 S18_1889 125.18 S18_1984 143.94 S18_3232 121.95 S18_1589 66.74 S18_3278 146.99 S18_3482 60.30 S18_3782 122.02 S18_4721 30.59 S24_2972 52.66 S24_3371 129.20 S24_3856 80.84 S24_4620 35.80 S32_2206 156.40 S18_1749 50.85 S18_2248 113.52 S18_2870 134.22 S18_3685 48.80 S24_1628 88.36 S18_1342 119.50 S18_2325 156.94 S18_2795 78.23 S18_4409 66.99 S18_4933 66.88 S24_1046 29.87 S24_1937 37.63 S24_2022 77.24 S24_2766 111.57 S24_2887 77.05 S24_3191 91.02 S24_3432 32.82 S24_3969 182.16 S10_1949 138.88 S10_4962 132.57 S12_1666 114.34 S18_1097 49.60 S18_1367 60.77 S18_2432 100.30 S18_2949 51.22 S18_2957 93.20 S18_3136 166.24 S12_1108 145.04 S12_3148 143.61 S12_3891 103.10 S12_4473 160.46 S18_2238 84.33 S18_3320 115.03 S18_4600 46.29 S18_4668 120.12 S24_2300 97.39 S24_4258 85.72 S32_1268 57.53 S32_3522 101.15 S700_2824 119.05 S18_2319 125.66 S18_3140 155.79 S18_3232 83.70 S18_3259 137.88 S18_4027 77.24 S18_4522 52.60 S24_1444 33.24 S24_2840 117.10 S24_4048 53.57 S32_2509 55.93 S32_3207 94.92 S50_1392 48.62 S50_1514 129.20 S10_4757 114.29 S24_2011 71.69 S24_3151 71.02 S700_1938 78.83 S24_3816 62.00 S700_1138 79.91 S32_1374 80.39 S10_1678 118.94 S10_2016 156.86 S10_4698 125.01 S12_2823 58.47 S32_4289 41.02 S50_1341 74.90 S700_1691 111.53 S700_2834 182.90 S12_1099 95.13 S12_3380 76.61 S12_3990 102.49 S12_4675 63.91 S18_1889 52.09 S18_2625 112.70 S24_1578 61.70 S24_2000 117.48 S18_1129 132.29 S18_1984 102.98 S18_2325 162.00 S18_2795 75.59 S18_3278 135.23 S18_3482 54.09 S18_3782 122.02 S18_4721 67.87 S24_2360 35.87 S24_2972 51.43 S24_3371 134.81 S24_3856 66.29 S24_4620 39.02 S32_2206 101.03 S32_4485 81.36 S50_4713 195.01 S10_1949 121.15 S10_4962 110.70 S12_1666 114.34 S18_1097 102.74 S18_1342 47.44 S18_1367 57.73 S18_2432 99.28 S18_2949 26.55 S24_1937 36.29 S24_2022 61.21 S18_2957 103.67 S18_3136 98.22 S18_3320 145.04 S12_3148 99.54 S12_4473 137.53 S18_2238 104.32 S18_2319 106.55 S18_4600 41.76 S18_4668 112.46 S24_2300 88.62 S24_4258 84.75 S32_1268 63.35 S32_3522 83.95 S700_2824 135.47 S18_3232 124.95 S18_4027 53.18 S24_1444 35.36 S24_2840 104.09 S24_4048 47.08 S32_2509 55.93 S32_3207 125.12 S10_4757 205.72 S12_1108 98.39 S50_1392 57.99 S50_1514 155.72 S12_3891 129.76 S18_3140 91.76 S18_3259 83.38 S18_4522 100.77 S24_2011 77.00 S24_31

In [0]:
df_q2 = df_orderdetails.join(df_orders_q2, df_orderdetails.order_number == df_orders_q2.order_number, 'INNER')\
                       .select(df_orderdetails.price_each, df_orderdetails.product_code)


df_q2 = df_q2.join(df_products, df_q2.product_code == df_products.product_code, 'INNER') 

In [0]:
df_q2 = df_q2.groupBy('product_line')\
             .agg(sum('price_each').alias('faturamento'))\
             .orderBy(col('faturamento').desc())\
             .limit(1)

In [0]:
df_q2.display()

product_line faturamento Classic Cars 15559.72

> <b>Resposta</b>
-  O faturamento da linha de produto mais vendido no ano de 2005 e com o status Shipped é $ 15.559,72

Salvando o resultado

In [0]:
df_q2.write.mode('overwrite').format('delta').option('mergeSchema', 'True').save('/join/gold/resposta_q2')

#### - Nome, sobrenome e email dos vendedores do Japão, o local-part do e-mail deve estar mascarado.

- Solução utilizando SQL

Dataframes abordados para essa questão:
- Employees
- Offices

In [0]:
df_employees = spark.read.format('delta').load('/join/silver/df_employees')
exibir_info_df(df_employees)

df_offices = spark.read.format('delta').load('/join/silver/df_offices')
exibir_info_df(df_offices)


Linhas = 23 
Colunas = 8

root
 |-- employee_number: integer (nullable = true)
 |-- last_name: string (nullable = true)
 |-- first_name: string (nullable = true)
 |-- extension: string (nullable = true)
 |-- email: string (nullable = true)
 |-- office_code: string (nullable = true)
 |-- reports_to: integer (nullable = true)
 |-- job_Title: string (nullable = true)


Linhas = 7 
Colunas = 9

root
 |-- office_code: string (nullable = true)
 |-- city: string (nullable = true)
 |-- phone: string (nullable = true)
 |-- address_line1: string (nullable = true)
 |-- address_line2: string (nullable = true)
 |-- state: string (nullable = true)
 |-- country: string (nullable = true)
 |-- postal_code: string (nullable = true)
 |-- territory: string (nullable = true)



In [0]:
df_employees.display()

employee_number last_name first_name extension email office_code reports_to job_Title 1002 Murphy Diane x5800 dmurphy@classicmodelcars.com 1 0 President 1056 Patterson Mary x4611 mpatterso@classicmodelcars.com 1 1002 VP Sales 1076 Firrelli Jeff x9273 jfirrelli@classicmodelcars.com 1 1002 VP Marketing 1088 Patterson William x4871 wpatterson@classicmodelcars.com 6 1056 Sales Manager (APAC) 1102 Bondur Gerard x5408 gbondur@classicmodelcars.com 4 1056 Sale Manager (EMEA) 1143 Bow Anthony x5428 abow@classicmodelcars.com 1 1056 Sales Manager (NA) 1165 Jennings Leslie x3291 ljennings@classicmodelcars.com 1 1143 Sales Rep 1166 Thompson Leslie x4065 lthompson@classicmodelcars.com 1 1143 Sales Rep 1188 Firrelli Julie x2173 jfirrelli@classicmodelcars.com 2 1143 Sales Rep 1216 Patterson Steve x4334 spatterson@classicmodelcars.com 2 1143 Sales Rep 1286 Tseng Foon Yue x2248 ftseng@classicmodelcars.com 3 1143 Sales Rep 1323 Vanauf George x4102 gvanauf@classicmodelcars.com 3 1143 Sales Rep 1337 Bondur Loui x6493 lbondur@classicmodelcars.com 4 1102 Sales Rep 1370 Hernandez Gerard x2028 ghernande@classicmodelcars.com 4 1102 Sales Rep 1401 Castillo Pamela x2759 pcastillo@classicmodelcars.com 4 1102 Sales Rep 1501 Bott Larry x2311 lbott@classicmodelcars.com 7 1102 Sales Rep 1504 Jones Barry x102 bjones@classicmodelcars.com 7 1102 Sales Rep 1611 Fixter Andy x101 afixter@classicmodelcars.com 6 1088 Sales Rep 1612 Marsh Peter x102 pmarsh@classicmodelcars.com 6 1088 Sales Rep 1619 King Tom x103 tking@classicmodelcars.com 6 1088 Sales Rep 1621 Nishi Mami x101 mnishi@classicmodelcars.com 5 1056 Sales Rep 1625 Kato Yoshimi x102 ykato@classicmodelcars.com 5 1621 Sales Rep 1702 Gerard Martin x2312 mgerard@classicmodelcars.com 4 1102 Sales Rep

In [0]:
df_offices.display()

office_code city phone address_line1 address_line2 state country postal_code territory 1 San Francisco +1 650 219 4782 100 Market Street Suite 300 CA USA 94080 NA 2 Boston +1 215 837 0825 1550 Court Place Suite 102 MA USA 02107 NA 3 NYC +1 212 555 3000 523 East 53rd Street apt. 5A NY USA 10022 NA 4 Paris +33 14 723 4404 43 Rue Jouffroy D'abbans Uninformed capital France 75017 EMEA 5 Tokyo +81 33 224 5000 4-1 Kioicho Uninformed Chiyoda-Ku Japan 102-8578 Japan 6 Sydney +61 2 9264 2451 5-11 Wentworth Avenue Floor #2 capital Australia NSW 2010 APAC 7 London +44 20 7877 2041 25 Old Broad Street Level 7 capital UK EC2N 1HN EMEA

In [0]:

df_office_japan = df_offices.filter('country = "Japan"')
df_office_japan.display()

office_code city phone address_line1 address_line2 state country postal_code territory 5 Tokyo +81 33 224 5000 4-1 Kioicho Uninformed Chiyoda-Ku Japan 102-8578 Japan

- Solução utilizando SQL

Criando tabelas temporárias

In [0]:
df_office_japan.createOrReplaceTempView('office')
df_employees.createOrReplaceTempView('emp')

Consulta final apresentando o resultado

In [0]:
%sql

SELECT 
  e.first_name,
  e.last_name,
  CONCAT('******', SUBSTRING(e.email, LOCATE('@', e.email))) AS email
FROM emp AS e
INNER JOIN office AS o 
ON e.office_code = o.office_code;

first_name last_name email Mami Nishi ******@classicmodelcars.com Yoshimi Kato ******@classicmodelcars.com

- Solução utilizando PySpark

In [0]:
df_employees.join(df_office_japan, df_employees.office_code == df_office_japan.office_code, 'INNER')\
           .select(df_employees.first_name, df_employees.last_name, df_employees.email).display()

first_name last_name email Mami Nishi mnishi@classicmodelcars.com Yoshimi Kato ykato@classicmodelcars.com

Consulta final apresentando o resultado

In [0]:
df_q3 = df_employees.join(df_office_japan, df_employees.office_code == df_office_japan.office_code, 'INNER')\
                    .select(df_employees.first_name, df_employees.last_name, df_employees.email)

df_q3 = df_q3.withColumn('email', regexp_replace('email', '[^@]+@', '*****@'))
df_q3.display()

first_name last_name email Mami Nishi *****@classicmodelcars.com Yoshimi Kato *****@classicmodelcars.com

Salvando o resultado

In [0]:
df_q3.write.mode('overwrite').format('delta').option('mergeSchema', 'True').save('/join/gold/resposta_q3')